### ISSUE TITLE
BFF token_exchange question

### ISSUE THREAD

**Issue by davidbzhealth** [DuendeSoftware/Support#965]
**Which version of Duende BFF are you using?**
1.2.0

**Which version of .NET are you using?**
6.0
 
We’re attempting to integrate with Cerner/FHir, our Duende Auth Server, and your BFF.  We’re working on a flow pattern where:
 
1.	A User starts at the Cerner site and Clicks on an application
2.	The user is then redirected to our site with a launchCode generated by the click
3.	Our site then redirects to  a code authorization request with the launch token to the cerner authorization server
4.	Then on their server the user selects a patient and logs in
5.	The Cerner site then redirects back to our site with a code and state in the query string
6.	We then use that code and state to get a cerner access_token
7.	With this authorized user we perform a token exchange through our Duende Auth Server to get our token
8.	[HERES WHERE I’M STUCK] from the BFF I’m trying to ‘login’ the user with the token we got back from our Duende Auth Server but 
1.	IUserAccessTokenStore. StoreTokenAsync seems to require the user already being logged in
2.	I don’t see any method in the BFF libraries to login in a user outside of the /bff/login flow

Am I fundamentally looking at this incorrectly? Should our BFF/Frontend be relying on the Single Sign-on flow and have /bff/login re-auth the user the typical code exchange flow?

Any help would be appreciated here,

**Comment by AndersAbel** [comment]
Yes, the BFF is designed to use an existing OpenID Connect flow to sign in with IdentityServer. The recommended flow to use is authorization code flow with PKCE.

The token exchange mechanism is designed for scenarios where a client has an access token and swaps it for another access token.

Session establishment is much more complicated from a security point of view. Their are multiple security measures that must be taken to prevent session pinning, code/token swap attacks and replay attacks.

I think that in your scenario the Cerner site should be treated as an upstream IdentityProvider to IdentityServer. By letting IdentityServer handle the handshake, your client application can stick to pure OpenID Connect in it's interaction with IdentityServer.

Going deeper into this is really outside of the scope of what we can do here on the public support tracker. If you want further assistance, we can provider that through our remote consulting services. Please reach out to contact@duendesoftware.com for more information.

**Comment by AndersAbel** [comment]
Are there any more questions about this or can we go ahead and close the issue?
